yarn-project/world-state/src/native/native_world_state.ts

```diff
@@ -0,0 +1,448 @@
+import { type L2Block, MerkleTreeId, SiblingPath, TxEffect } from '@aztec/circuit-types';
+import {
+  AppendOnlyTreeSnapshot,
+  ContentCommitment,
+  Fr,
+  GlobalVariables,
+  Header,
+  MAX_NOTE_HASHES_PER_TX,
+  MAX_NULLIFIERS_PER_TX,
+  MAX_TOTAL_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX,
+  NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP,
+  NullifierLeaf,
+  NullifierLeafPreimage,
+  PartialStateReference,
+  PublicDataTreeLeaf,
+  PublicDataTreeLeafPreimage,
+  StateReference,
+} from '@aztec/circuits.js';
+import { padArrayEnd } from '@aztec/foundation/collection';
+import { SerialQueue } from '@aztec/foundation/fifo';
+import { serializeToBuffer } from '@aztec/foundation/serialize';
+import type { IndexedTreeLeafPreimage } from '@aztec/foundation/trees';
+import type { BatchInsertionResult } from '@aztec/merkle-tree';
+
+import bindings from 'bindings';
+import { Decoder, Encoder, addExtension } from 'msgpackr';
+import { isAnyArrayBuffer } from 'util/types';
+
+import { type MerkleTreeDb, type TreeSnapshots } from '../world-state-db/merkle_tree_db.js';
+import {
+  type HandleL2BlockAndMessagesResult,
+  type IndexedTreeId,
+  type MerkleTreeLeafType,
+  MerkleTreeOperations,
+  type TreeInfo,
+} from '../world-state-db/merkle_tree_operations.js';
+import { MerkleTreeOperationsFacade } from '../world-state-db/merkle_tree_operations_facade.js';
+import {
+  MessageHeader,
+  SerializedIndexedLeaf,
+  SerializedLeafValue,
+  TypedMessage,
+  WorldStateMessageType,
+  type WorldStateRequest,
+  type WorldStateResponse,
+  blockStateReference,
+  treeStateReferenceToSnapshot,
+  worldStateRevision,
+} from './message.js';
+
+// small extension to pack an NodeJS Fr instance to a representation that the C++ code can understand
+// this only works for writes. Unpacking from C++ can't create Fr instances because the data is passed
+// as raw, untagged, buffers. On the NodeJS side we don't know what the buffer represents
+// Adding a tag would be a solution, but it would have to be done on both sides and it's unclear where else
+// C++ fr instances are sent/received/stored.
+addExtension({
+  Class: Fr,
+  write: fr => fr.toBuffer(),
+});
+
+export interface NativeInstance {
+  call(msg: Buffer | Uint8Array): Promise<any>;
+}
+
+export class NativeWorldStateService implements MerkleTreeDb {
+  private nextMessageId = 1;
+
+  private encoder = new Encoder({
+    // always encode JS objects as MessagePack maps
+    // this makes it compatible with other MessagePack decoders
+    useRecords: false,
+    int64AsType: 'bigint',
+  });
+
+  private decoder = new Decoder({
+    useRecords: false,
+    int64AsType: 'bigint',
+  });
+
+  private queue = new SerialQueue();
```

Comment:
Serialize all writes and reads to the cpp code. Ideally this would only lock for uncomitted state.